File: question/sort/src/src/sort.cpp
#include <iostream>
#include <vector>

using namespace std;


/**
 * @brief 包含各种排序算法
 * 
 */
class Sort
{
public:
    vector<int> arr;
    int size;

    Sort(vector<int>& arr, int size){
        this->arr = arr;
        this->size = size;
    }
    
    void PrintArr(){
        for (int i = 0; i < this->size; i++) {
            cout << this->arr[i] << " ";
        }
        cout << endl;
    }
    
    // ****************************** quickSort ****************************** //
    int partition(int low, int high) {
        int pivot = arr[high];  // 将最后一个元素选为基准值
        int i = low - 1;
        
        for (int j = low; j <= high - 1; j++) {
            if (arr[j] <= pivot) {
                i++;
                swap(arr[i], arr[j]);  // 将较小的元素交换到前面
            }
        }
        
        swap(arr[i + 1], arr[high]);  // 将基准值放置到正确的位置
        return (i + 1);
    }

    void quickSortMain(int low, int high) {
        if (low < high) {
            int pivotIndex = partition(low, high);      // 划分数组
            // cout << "pivotIndex: " << pivotIndex << " low: " << low << " high: "  << high << endl;
            quickSortMain(low, pivotIndex - 1);         // 递归左子数组排序
            quickSortMain(pivotIndex + 1, high);        // 递归右子数组排序
        }
    }

    void quickSort(){
        quickSortMain(0, this->size - 1);
    }

    // ****************************** mergeSort ****************************** //
    
    // ~Sort();
};


int main() {
    vector<int> arr = {3, 1, 8, 4, 2};
    int n = arr.size();

    Sort sort(arr, n);
    sort.quickSort(); 
    sort.PrintArr();

    return 0;
}

File: question/sort/script/sort.py
# coding=utf-8

class Sort:

    def __init__(self, arr: list):
        self.arr = arr
        self.length = len(self.arr)

    def swap(self, num1, num2):
        tmp = num1
        num1 = num2
        num2 = tmp
        return num1, num2

    def bubble_sort(self):
        for i in range(self.length-1):
            for j in range(0, self.length-1-i):
                if self.arr[j] > self.arr[j+1]:
                    self.arr[j], self.arr[j+1] = self.swap(self.arr[j], self.arr[j+1])
        return self.arr

    def insert_sort(self):
        for i in range(1, self.length):
            base = self.arr[i]
            j = i-1
            while j >= 0 and self.arr[j] > base:
                self.arr[j+1] = self.arr[j]
                j -= 1
            self.arr[j+1] = base
        return self.arr

    def selection_sort(self):
        """
        O(n)=O(n^2), T(n)=O(1)
        :param arr:
        :return:
        """
        for i in range(self.length):
            for j in range(i+1, self.length):
                if self.arr[i] > self.arr[j]:
                    tmp = self.arr[i]
                    self.arr[i] = self.arr[j]
                    self.arr[j] = tmp
        return self.arr

    def quick_sort_main(self, start, end):
        if start >= end:
            return
        mid_data, left, right = self.arr[start], start, end
        while left < right:
            while self.arr[right] >= mid_data and left < right:
                right -= 1
            self.arr[left] = self.arr[right]
            while self.arr[left] < mid_data and left < right:
                left += 1
            self.arr[right] = self.arr[left]
        self.arr[left] = mid_data
        self.quick_sort_main(start, left - 1)
        self.quick_sort_main(left + 1, end)
        return self.arr

    def quick_sort(self):
        return self.quick_sort_main(0, self.length-1)

    def quick_sort_plus_main(self, lst):
        if len(lst) <= 1:
            return lst
        else:
            pivot = lst[0]  # 选择第一个元素作为基准值
            less = [x for x in lst[1:] if x <= pivot]  # 小于等于基准值的子数组
            greater = [x for x in lst[1:] if x > pivot]  # 大于基准值的子数组
            return self.quick_sort_plus_main(less) + [pivot] + self.quick_sort_plus_main(greater)

    def quick_sort_plus(self):
        return self.quick_sort_plus_main(self.arr)
        

if __name__ == '__main__':
    array = [10, 17, 50, 7, 30, 30, 27, 45, 15, 5, 36, 21]

    sort = Sort(array)
    # print(sort.bubble_sort())
    print(sort.insert_sort())
    # print(sort.quick_sort())
    # print(sort.quick_sort_plus())
    # print(sort.selection_sort())